File: scopes.js
for (let i = 0; i <10; i++) {
    console.log(i);
}

// si la variable i n'a pas été définie dans le scope global, elle n'est pas utilisable
// si la variable i à été définie avec le mot clé let, elle n'est accessible que 
//si elle a été définie dans le même scope (ou un scope extérieur)
console.log(i);

for (var i = 0; i <10; i++) {
    console.log(i);
}

// si la variable i à été définie avec le mot clé var, elle est accessible même si elle a été définie dans un scope intérieur
console.log(i);

function myFunc1() {
    var myVar1 = 123;
}

myFunc1();
// même si la variable myVar1 est déclarée avec le mot clé var, comme elle est déclarée à l'intérieur d'une fonction, 
// elle n'est pas accessible depuis un scope extérieur
// ReferenceError: myVar1 is not defined
console.log(myVar1);

File: scopes.py
# scope = portée des variables

my_var1 = 123

def my_func1():
    print(my_var1)
    print(my_var2)

my_var2 = 42

# la fonction voit les variables définies à l'extérieur au préalable ou à postériorie
my_func1()

def my_func2():
    my_var3 = 3.14

my_func2()

# il n'est pas possible d'accéder à une variable définie à l'intérieur d'une fonction
# que celle-ci ait été exécutée ou non
# principe d'un verre teinté.
# NameError: name 'my_var3' is not defined.
# print(my_var3)

my_var4 = 'foo'

def my_func3(my_var4):
    # le paramètre my_var4 masque la variable définie à l'extérieur de la fonction
    print(my_var4)

# cet appel affiche 'bar'
my_func3('bar')

my_var4 = 'foo'

def my_func4():
    # la varible my_var4 fait de l'ombre (shadowing) à la variable définie à l'extérieur de la fonction
    my_var4 = 'baz'
    print(my_var4)

my_func4()
# la variable définie à l'extérieur de la fonction reste inchangée
print(my_var4)
